### PR TITLE
track rebalance status when instance assignment already correct

### PR DESCRIPTION
**Pull request by jadami10**
This is a small `bugfix` for the table rebalancer. This triggers the start of the rebalance observer earlier. It then sets it to success when instance assignment is already complete. If you rebalance a table with "reassignAssignments" and the instance assignment doesn't change, you will get "IN_PROGRESS" back, but the id won't exist in zk. So polling that id will just give you a 404, but the logs will say the table is already balanced.

After this fix, the id will get updated in ZK, so you can poll it. I updated existing unit tests, and I also tested with the hybrid quickstart. Though testing with code from 5 months ago vs latest master seems to get different results. The older code would return "IN_PROGRESS" consistently, and now I'm getting back "NO_OP" which is also correct. I think #11073 actually fixed this, but there's likely still some cases where this helps.

### CODE REVIEW

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/rebalance/ZkBasedTableRebalanceObserver.java [hunk 1]
```diff
@@ -98,7 +98,8 @@ private void updateOnStart(Map<String, Map<String, String>> currentState,
 
   @Override
   public void onSuccess(String msg) {
-    Preconditions.checkState(_tableRebalanceProgressStats.getStatus() != RebalanceResult.Status.DONE.toString(),
+    Preconditions.checkState(
```
Comment (jadami10) [inline pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/rebalance/ZkBasedTableRebalanceObserver.java:101]: intellij was complaining that this comparison was not correct. but i'm happy to undo this change if we don't want to touch this

File: pinot-controller/src/test/java/org/apache/pinot/controller/helix/core/rebalance/TableRebalancerClusterStatelessTest.java [hunk 2]
```diff
@@ -66,13 +74,16 @@ public class TableRebalancerClusterStatelessTest extends ControllerTest {
   private static final String TIER_A_NAME = "tierA";
   private static final String TIER_B_NAME = "tierB";
   private static final String TIER_FIXED_NAME = "tierFixed";
+  private static final TableRebalanceObserver TABLE_REBALANCE_OBSERVER = mock(TableRebalanceObserver.class);
```
Comment (jadami10) [inline pinot-controller/src/test/java/org/apache/pinot/controller/helix/core/rebalance/TableRebalancerClusterStatelessTest.java:77]: this class was already testing a bunch of cases, so adding a mock here to check the observer seemed the best way to test.

**Comment by Jackie-Jiang** [comment]
Currently we only track the rebalance status when it returns `IN_PROGRESS` (execute in background), and not track the rebalance status when it finishes immediately (`NO_OP` or `DONE`). This PR changes that to always track the rebalance status even if it returns immediately.
Do we return the job id when the job finishes immediately (IMO that is not really needed)? If we want to track all the jobs, we also need to update the status for dry-run etc (basically any exit of the rebalance).

**Comment by Jackie-Jiang** [comment]
cc @klsince to also take a look

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/rebalance/TableRebalancer.java [hunk 3]
```diff
@@ -241,13 +241,19 @@ public RebalanceResult rebalance(TableConfig tableConfig, RebalanceConfig rebala
...
     boolean segmentAssignmentUnchanged = currentAssignment.equals(targetAssignment);
     LOGGER.info("For rebalanceId: {}, instancePartitionsUnchanged: {}, tierInstancePartitionsUnchanged: {}, "
             + "segmentAssignmentUnchanged: {} for table: {}", rebalanceJobId, instancePartitionsUnchanged,
         tierInstancePartitionsUnchanged, segmentAssignmentUnchanged, tableNameWithType);
 
     if (segmentAssignmentUnchanged) {
-      LOGGER.info("Table: {} is already balanced", tableNameWithType);
+      String msg = String.format("Table: %s is already balanced", tableNameWithType);
+      LOGGER.info(msg);
+      _tableRebalanceObserver.onSuccess(msg);
```
Comment (klsince) [inline pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/rebalance/TableRebalancer.java:256]: > If you rebalance a table with "reassignAssignments" and the instance assignment doesn't change, you will get "IN_PROGRESS" back, but the id won't exist in zk 

I'd assume we got a DONE status at L255 `if (dryRun) {...}` firstly (because rebalance restful API always does a dryRun firstly and triggers real trigger if dryRun gives DONE). But likely, the real table rebalance must have ended at the L259 `else {...}`. Because the observer.onTrigger(...START_TRIGGER) was run much later, so no job status got put in ZK for the jobId.

If that's the case, then I'm wondering why we didn't return NO_NP for all three if branches below? cc @Jackie-Jiang

Comment (klsince) [reply pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/rebalance/TableRebalancer.java:256]: and in general, I'd prefer we just track job status in `ZK` for table rebalances that really move segments around and potentially take time to finish, so returning NO_OP for all cases of 'table is already balanced'
